electricitymap/contrib/libs/models/datapoints.py

```diff
@@ -0,0 +1,427 @@
+from abc import ABC, abstractmethod
+from datetime import datetime, timedelta, timezone
+from logging import Logger
+from typing import Any, Dict, List, Optional
+
+from pydantic import BaseModel, validator
+
+from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG, ZoneKey
+from electricitymap.contrib.libs.models.constants import VALID_CURRENCIES
+
+LOWER_DATETIME_BOUND = datetime(2000, 1, 1, tzinfo=timezone.utc)
+
+
+class ProductionMix(BaseModel):
+    """
+    Containts the production mix for a zone at a given time.
+    All values are in MW.
+    """
+
+    biomass: Optional[float] = None
+    coal: Optional[float] = None
+    gas: Optional[float] = None
+    geothermal: Optional[float] = None
+    hydro: Optional[float] = None
+    nuclear: Optional[float] = None
+    oil: Optional[float] = None
+    solar: Optional[float] = None
+    unknown: Optional[float] = None
+    wind: Optional[float] = None
+
+    def set_value(self, mode: str, value: float) -> None:
+        """
+        Sets the value of a production mode.
+        This can be used if the Production has been initialized empty
+        and is being filled in a loop.
+        """
+        self.__setattr__(mode, value)
+
+
+class StorageMix(BaseModel):
+    """
+    Containts the storage mix for a zone at a given time.
+    All values are in MW.
+    """
+
+    battery: Optional[float] = None
+    hydro: Optional[float] = None
+
+
+class Event(BaseModel, ABC):
+    """
+    An abstract class representing all types of electricity events that can occur in a zone.
+    forecasted: Whether the point is a forecasted point or not.
+    Should be set to True if the point is a forecast provided by a datasource.
+    zoneKey: The zone key of the zone the event is happening in.
+    datetime: The datetime of the event.
+    source: The source of the event.
+    We currentlu use the root url of the datasource. Ex: edf.fr
+    """
+
+    forecasted: bool = False
+    zoneKey: ZoneKey
+    datetime: datetime
+    source: str
+    # TODO estimated: bool = False,
+
+    @validator("zoneKey")
+    def _validate_zone_key(cls, v):
+        if v not in ZONES_CONFIG:
+            raise ValueError(f"Unknown zone: {v}")
+        return v
+
+    @validator("datetime")
+    def _validate_datetime(cls, v, values: Dict[str, Any]):
+        if v.tzinfo is None:
+            raise ValueError(f"Missing timezone: {v}")
+        if v < LOWER_DATETIME_BOUND:
+            raise ValueError(f"Date is before 2000, this is not plausible: {v}")
+        if not values.get("forecasted", False) and v > datetime.now(
+            timezone.utc
+        ) + timedelta(days=1):
```

Comment:
Adding a delta of one day to add a margin of error. We have seen some flaky test in the past with such procedures.